### PR TITLE
T248354 update useI18n() interface

### PR DESCRIPTION
**Pull request by hueitan**
Following the PR of https://github.com/wikimedia/wikipedia-kaios/pull/190

Phabricator Link: https://phabricator.wikimedia.org/T248354

### Problem Statement

banana doesn't allow us to grab another locale translation without changing the locale

### Solution

Two interface usage of i18n : `i18n()` for UI lang, `i18n( lang )` for given lang translation

simplify `i18n.i18n()` to just `i18n()`

### Note

<s>Test spec completely fail, but in the meanwhile, this is ready to be reviewed</s>

### CODE REVIEW

**Comment by hueitan** [comment]
Oppss,, I was supposed to open it as Draft PR 😮

File: src/hooks/useI18n.js [hunk 1]
```diff
@@ -1,4 +1,19 @@
 import { useContext } from 'preact/hooks'
 import { I18nContext } from 'contexts'
+import { getAppLanguage } from 'utils'
 
-export const useI18n = () => useContext(I18nContext)
+export const useI18n = lang => {
+  const banana = useContext(I18nContext)
+  banana.setLocale(getAppLanguage())
```
Comment (hueitan) [inline src/hooks/useI18n.js]: do we foresee the performance issue when reseting the locale everytime it calls the useI18n?

Comment (stephanebisson) [reply src/hooks/useI18n.js]: Yes, let's not do that.

https://github.com/wikimedia/banana-i18n/blob/7259156f439fefe9e1239888585ecec6a02699ff/src/index.js#L31

File: src/hooks/useI18n.js [hunk 2]
```diff
@@ -1,4 +1,19 @@
...
+
+  return (key, ...args) => {
+    if (lang) {
+      const locale = banana.locale
+      banana.setLocale(lang)
+      const msg = banana.i18n(key, args)
```
Comment (stephanebisson) [inline src/hooks/useI18n.js]: Do you need to spread `args` like you do below?

File: src/hooks/useI18n.js [hunk 3]
```diff
@@ -1,4 +1,19 @@
...
+import { getAppLanguage } from 'utils'
 
-export const useI18n = () => useContext(I18nContext)
+export const useI18n = lang => {
+  const banana = useContext(I18nContext)
+  banana.setLocale(getAppLanguage())
+
+  return (key, ...args) => {
+    if (lang) {
+      const locale = banana.locale
```
Comment (stephanebisson) [inline src/hooks/useI18n.js]: This logic should be in a try/finally block to ensure we leave the instance with the right language.

File: src/hooks/useI18n.js (hunk 1 above)
Comment (hueitan) [reply src/hooks/useI18n.js]: Update in https://github.com/wikimedia/wikipedia-kaios/pull/196/commits/caa953a7f27a135d9b3cbfa777a8f074a0f5b3ee

It's still needed but only execute when changing the app language

File: src/hooks/useI18n.js [hunk 4]
```diff
@@ -1,4 +1,30 @@
...
+  const banana = useContext(I18nContext)
+
+  return (key, ...args) => {
+    if (lang) {
+      const locale = banana.locale
+      let msg
+      try {
+        banana.setLocale(lang)
+      } finally {
+        msg = banana.i18n(key, ...args)
```
Comment (stephanebisson) [inline src/hooks/useI18n.js]: I think this line should be in the `try` block

Comment (hueitan) [reply src/hooks/useI18n.js]: if this line in `try` block, msg will be undefined, we need to get it as least the fallback lang

Comment (stephanebisson) [reply src/hooks/useI18n.js]: This is not related to fallback. The try/finally is to ensure that if `banana.i18n(...)` throws an exception, the locale in the shared instance is reset to what it should be.

However, re-reading the code, I think the logic can be simplified to something like
```
lang = lang || getAppLanguage()
if (banana.locale !== lang) {
  banana.setLocale(lang)
}
return return banana.i18n(key, ...args)
```

The idea is that every time the function is called, it sets the right locale (if it needs to). Do you see any issue with this approach?

Comment (hueitan) [reply src/hooks/useI18n.js]: it totally makes sense and the code much simpler now

what benefit actually do we get in try/catch form when setLocale? identify the language we are not supported yet?

**Comment by stephanebisson** [comment]
LGTM

I'll merge it after you update the PR description.

**Comment by hueitan** [comment]
Test specs are fixed and ready to be reviewed / tested / merged.

**Comment by medied** [comment]
I'm kind of late to the party here but just reviewed this and wow this is great, I really like this refactor 👍

**Pull request merged**